### DISCUSSION TITLE
How to properly setup an SQLite Data Source?

### DISCUSSION THREAD

**Discussion by hugoaboud** [getredash/redash#6139] [Q&A]
I've deployed an AWS instance of Redash from the AMI and updated it to `10.1.0`, however I wasn't able to find any instructions on how to setup the SQLite integration.

What I've tried:
- Uploaded the SQLite database to the `server` docker, at `/app/mydata.db`
- Created a `SQLite Data Source` with `Database Path: /app/mydata.db`
- Clicked "Test Connection", the `/api/data_sources/2/test` request returned `200 OK`
- Navigated to "Create New Query"
- The data source schema flashes briefly, then disappears, and any query will fail
- Upon inspection, I've found that the `/api/data_sources/2/schema` request does return `200 OK`, however with an empty `schema` as response
![image](https://github.com/getredash/redash/assets/2349393/0210bee3-0190-4957-aa68-f7c6d031055b)
- Changing the `Database Path` and clicking `Test Connection` again eventually returns `200 OK` for any path.

What am I missing?

**Comment by justinclift** [comment]
That's a really good question.  I personally do a bunch of stuff with SQLite, but don't remember ever trying to directly upload a SQLite database to the backend containers.

Hopefully someone else knows how to get this done fairly easily. :smile:

---

Just to point out, when those Redash containers are recreated that database you uploaded will be blown away.  So, hopefully you have backups. :smile:

You could probably fix that by using a "volume: " entry in the docker-compose.yml file, but lets leave that for some other time...

**Comment by hugoaboud** [comment]
After some debugging I've found that:

- The docker container which actually reads the database file is the `adhoc_worker`
- The best way to make sure all containers reach it is by using a `volume` at `/opt/redash/docker-compose.yaml`, and then restarting the containers with `docker-compose restart`:
```
version: "2"
x-redash-service: &redash-service
  image: redash/redash:10.1.0.b50633
  depends_on:
    - postgres
    - redis
  env_file: /opt/redash/env
  restart: always
  volumes:
    - /home/ubuntu/sqlite:/sqlite
```
  - On the scenario above, you should use the `/sqlite/mydata.db` path on Redash.
  - You should put the `mydata.db` file on the host `/home/ubuntu/sqlite` folder, to avoid it being wiped when the containers are recreated.

About the surrounding bugs:

- SQLite Test Connection is broken because `sqlite3` creates a file if it doesn't exist, so it always returns that a file was found, except when trying on directories where the `redash` user can't write, such as `/`.
  - Here's a very simple suggestion on how to fix it, with the proper guidance I can open a PR for that: https://stackoverflow.com/questions/57828286/how-to-make-sqlite3-connect-fail-if-the-db-file-does-not-exist

- Even though the query returns results, the schema returned is empty: 
![image](https://github.com/getredash/redash/assets/2349393/bd4abc5e-f521-4f17-8110-b3ec7965bc71)

  - I couldn't figure this one out. The `Sqlite._get_tables` method seems to return the expected value.

**Reply by justinclift** [reply]
> The best way to make sure all containers reach it is by using a volume ...

Cool.  That sounds like a good approach. :smile:

For the rest, I'm not sure.  Hopefully someone else has time to look at this. :smile:

**Reply by davecahill** [reply]
This was really helpful for me @hugoaboud , thanks!
I had the sqlite volume mounted on the server pod, switching it to the workers one did the trick.

**Comment by hugoaboud** [comment]
Cool, closing this for now. Thanks for the support!
